File: ml-benchmark/fusion/algprog/page62.f
Inductive List = nil Unit | cons {Int, List};
Config NonLinear = true;

incre = \x: Int.
  let w = / x 10 in {w, - x (* 10 w)};

Number = {Int, List};

double = \n: Number.
  let current = fix (
    \f: List -> {Int, List}. \xs: List.
    match xs with
      nil _ -> {0, nil unit}
    | cons {h, t} ->
      let subres = f t in
        let info = incre (+ h (+ h subres.1)) in
          {info.1, cons {info.2, subres.2}}
    end
  ) in let info = current n.2 in
    {+ n.1 (+ n.1 info.1), info.2};

dnum = 10;
ratio = (fix (
  \f: Int -> Int. \n: Int.
  if == n 0 then 1 else * 2 (f (- n 1))
)) dnum;

head = \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> h
end;

multi = \xs: List. fix (
  \f: Int -> Number. \n: Int.
  if <= n 0 then {0, xs}
  else double (f (- n 1))
) dnum;

round = \xs: List.
 let res = multi xs in
  if >= (head res.2) 5 then + 1 (res.1)
  else res.1;

repr = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> cons {h, f t}
  end
);

val = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} ->
    if or (< h 0) (>= h 10) then false
    else f t
  end
);

main = \xs: List. if val xs then round (repr xs) else 0;

File: ml-benchmark/fusion/algprog/page62.ml
type l = Nil | Cons of int * l

let incre x =
  let w = x / 10 in
  (w, x - (10 * w))

let double n =
  let rec current xs =
    match xs with
    | Nil -> (0, Nil)
    | Cons (h, t) ->
        let subres = current t in
        match subres with
        | (sr1, sr2) ->
            let a = h + (h + sr1) in
            let info = incre a in
            match info with
            | (i1, i2) -> (i1, Cons (i2, sr2))
  in
  match n with
  | (n1, n2) ->
      let info = current n2 in
      match info with
      | (i1, i2) -> (n1 + (n1 + i1), i2)

let dnum = 10

let rec ratio_f n =
  if n = 0 then 1 else 2 * ratio_f (n - 1)

let ratio = ratio_f dnum

let head xs =
  match xs with
  | Nil -> 0
  | Cons (h, t) -> h

let multi xs =
  let rec f n =
    if n <= 0 then (0, xs)
    else
      let res = f (n - 1) in
      double res
  in
  f dnum

let round xs =
  let res = multi xs in
  match res with
  | (r1, r2) ->
      if head r2 >= 5 then r1 + 1 else r1

let rec repr xs =
  match xs with
  | Nil -> Nil
  | Cons (h, t) -> Cons (h, repr t)

let rec valid xs =
  match xs with
  | Nil -> true
  | Cons (h, t) ->
      if h < 0 || h >= 10 then false else valid t

let program xs =
  if valid xs then round (repr xs) else 0